gcc/rust/typecheck/rust-hir-type-check-expr.h

```diff
@@ -1146,6 +1147,8 @@ class TypeCheckExpr : public TypeCheckBase
 		rust_error_at (seg.get_locus (), "reverse lookup failure");
 		return new TyTy::ErrorType (expr.get_mappings ().get_hirid ());
 	      }
+            // MARC: why is this not an error to have a NodeId not referering to any HIR?
+            // Is it because we may be using a to-be-completed HIR graph ?
 	    return root_tyty;
```

Comment:
The reason it is not an error is consider consider a test case:

```rust
struct <T>Foo(T);

impl Foo<i32> {
  fn test() ...
}

impl Foo<f32> {
  fn test() ...
}

fn main() {
  let a;
  a = Foo::new();  
}
```

In that path of Foo::new we must take into consideration any generic arguments to determine which new function we should be calling to. So this is where the Path Probe takes place over in: https://github.com/Rust-GCC/gccrs/blob/7d3157eec84ef8620f4648200f54b293e561b8c4/gcc/rust/typecheck/rust-hir-path-probe.h#L42

This means we take the type that has been resolved for the segments up to this point and use it as the receiver to match an impl block for that type and then look for any impl items which match the segment we want.

Reply:
Ok, thanks, I think I understand a bit better now!